extapi.go

```diff
@@ -0,0 +1,121 @@
+package rcmgr
+
+import (
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+// ResourceScopeLimiter is a trait interface that allows you to access scope limits.
+type ResourceScopeLimiter interface {
+	Limit() Limit
+	SetLimit(Limit)
+}
+
+var _ ResourceScopeLimiter = (*resourceScope)(nil)
+
+// ResourceManagerStat is a trait that allows you to access resource manager state.
+type ResourceManagerState interface {
+	ListServices() []string
+	ListProtocols() []protocol.ID
+	ListPeers() []peer.ID
+
+	Stat() ResourceManagerStat
+}
+
+type ResourceManagerStat struct {
+	System    network.ScopeStat
+	Transient network.ScopeStat
+	Services  map[string]network.ScopeStat
+	Protocols map[protocol.ID]network.ScopeStat
+	Peers     map[peer.ID]network.ScopeStat
+}
+
+var _ ResourceManagerState = (*resourceManager)(nil)
+
+func (s *resourceScope) Limit() Limit {
+	s.Lock()
+	defer s.Unlock()
+
+	return s.rc.limit
+}
+
+func (s *resourceScope) SetLimit(limit Limit) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.rc.limit = limit
+}
+
+func (r *resourceManager) ListServices() []string {
```

Comment:
Should we sort the result in all `List*` functions? Map iterations are (pseudo-) random.

Reply:
we could i guess, why not.

Reply:
done.